### PR TITLE
feat(s3-presigned-post): support S3 object post

### PR DESCRIPTION
**Pull request by AllanFly120**
Add support for generating presigned post url and fields. Users can use them to upload file without AWS credentials. 

Resolves: #395 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/s3-presigned-post/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+You can optionally attach a policy to a presigned post. It specifies a list of conditions that the request must meet.
+For example:
+
+```typescript
+const Conditions = [{ acl: "public-read" }, { bucket: "johnsmith" }, ["starts-with", "$key", "user/eric/"]];
+```
+
+More supported policy elements are documented in [S3 POST documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html).
+Note that if you include a condition, you must specify the valid value in the `Fields` parameter as well. A value will
+not be added automatically to the fields dictionary according to the conditions.
```
Comment (trivikr) [inline packages/s3-presigned-post/README.md:36]: Switched the active voice, and removed "note that"
```suggestion
Visit [S3 POST documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html) for supported policy elements.
If you include a condition, you must specify the valid value in the `Fields` parameter as well. A value will
not be added automatically to the fields dictionary according to the conditions.
```

File: packages/s3-presigned-post/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      <input type="file" name="file" /> <br />
+      <input type="submit" name="submit" value="Upload to Amazon S3" />
+    </form>
+  </body>
+</html>
+```
+
+### Post File using FormData in Node.js
+
+You can use Node.js `request` module to post a file as bellow:
```
Comment (trivikr) [inline packages/s3-presigned-post/README.md:89]:
```suggestion
In Node.js, use `request` module to post a file:
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  Expires?: number;
+}
+
+export interface PresignedPost {
+  url: string;
+  fields: {};
+}
+
+/**
+ * Builds the url and the form fields used for a presigned s3 post
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:33]:
```suggestion
 * Builds the url and the form fields used for a presigned s3 post.
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  const fields: Fields = {
+    ...Fields,
+    bucket: Bucket,
+    [ALGORITHM_QUERY_PARAM]: ALGORITHM_IDENTIFIER,
+    [CREDENTIAL_QUERY_PARAM]: credential,
+    [AMZ_DATE_QUERY_PARAM]: signingDate,
+    ...(clientCredentials.sessionToken ? { [TOKEN_QUERY_PARAM]: clientCredentials.sessionToken } : {}),
+  };
+
+  // prepare policies
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:59]:
```suggestion
  // Prepare policies.
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  ];
+  const encodedPolicy = base64Encoder(
+    utf8Decoder(
+      JSON.stringify({
+        expiration: iso8601(expiration),
+        conditions,
+      })
+    )
+  );
+  // sign the request
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:76]:
```suggestion

  // Sign the request.
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  const signingKey = await getSigningKey(sha256, clientCredentials, shortDate, clientRegion, "s3");
+  const signature = await hmac(sha256, signingKey, encodedPolicy);
+
+  const endpoint = await client.config.endpoint();
+  if (!client.config.bucketEndpoint) {
+    endpoint.path = `/${Bucket}`;
+  }
+  return {
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:84]:
```suggestion

  return {
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * Builds the url and the form fields used for a presigned s3 post
+ */
+export const createPresignedPost = async (
+  client: S3Client,
+  { Bucket, Key, Conditions = [], Fields = {}, Expires = 3600 }: PresignedPostOptions
+): Promise<PresignedPost> => {
+  const { systemClockOffset, base64Encoder, utf8Decoder, sha256 } = client.config;
+  const now = new Date(Date.now() + systemClockOffset);
+  // signingDate in format like '20201028T070711Z'
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:41]:
```suggestion

  // signingDate in format like '20201028T070711Z'.
```

File: packages/s3-presigned-post/src/createPresignedPost.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  const signingDate = iso8601(now).replace(/[\-:]/g, "");
+  const shortDate = signingDate.substr(0, 8);
+  const clientRegion = await client.config.region();
+  // prepare credentials
```
Comment (trivikr) [inline packages/s3-presigned-post/src/createPresignedPost.ts:45]:
```suggestion

  // Prepare credentials.
```

**Review by trivikr — APPROVED** [review]

**Comment by AllanFly120** [comment]
Addressed the documenting & formatting feedbacks in 5f45ba0, with some other changes in README.

**Pull request merged**